x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionEncoder.java

```diff
@@ -0,0 +1,220 @@
+    public static final byte PRERELESE_SEPARATOR_BYTE = (byte) 0x02;
+    public static final byte NO_PRERELESE_SEPARATOR_BYTE = (byte) 0x03;
+
+    private static final char PRERELESE_SEPARATOR = '-';
+    private static final char DOT_SEPARATOR = '.';
+    private static final char BUILD_SEPARATOR = '+';
+
+    // Regex to test relaxed Semver Main Version validity. Allows for more or less than three main version parts
+    private static Pattern LEGAL_MAIN_VERSION_SEMVER = Pattern.compile("(0|[1-9]\\d*)(\\.(0|[1-9]\\d*))*");
+
+    private static Pattern LEGAL_PRERELEASE_VERSION_SEMVER = Pattern.compile(
+        "(?:-((?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\\.(?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*))*))"
+    );
+
+    private static Pattern LEGAL_BUILDSUFFIX_SEMVER = Pattern.compile("(?:\\+([0-9a-zA-Z-]+(?:\\.[0-9a-zA-Z-]+)*))?");
+
+    /**
+     * Encodes a version string.
+     */
+    public static EncodedVersion encodeVersion(String versionString) {
+        VersionParts versionParts = VersionParts.ofVersion(versionString);
+
+        // don't treat non-legal versions further, just mark them as illegal and return
+        if (legalVersionString(versionParts) == false) {
+            return new EncodedVersion(new BytesRef(versionString), false, true, 0, 0, 0);
+        }
+
+        BytesRefBuilder encodedBytes = new BytesRefBuilder();
+        Integer[] mainVersionParts = prefixDigitGroupsWithLength(versionParts.mainVersion, encodedBytes);
+
+        if (versionParts.preRelease != null) {
+            encodedBytes.append(PRERELESE_SEPARATOR_BYTE);  // versions with pre-release part sort before ones without
+            encodedBytes.append((byte) PRERELESE_SEPARATOR);
+            String[] preReleaseParts = versionParts.preRelease.substring(1).split("\\.");
+            boolean first = true;
+            for (String preReleasePart : preReleaseParts) {
+                if (first == false) {
+                    encodedBytes.append((byte) DOT_SEPARATOR);
+                }
+                boolean isNumeric = preReleasePart.chars().allMatch(x -> Character.isDigit(x));
+                if (isNumeric) {
+                    prefixDigitGroupsWithLength(preReleasePart, encodedBytes);
+                } else {
+                    encodedBytes.append(new BytesRef(preReleasePart));
+                }
+                first = false;
+            }
+        } else {
+            encodedBytes.append(NO_PRERELESE_SEPARATOR_BYTE);
+        }
+
+        if (versionParts.buildSuffix != null) {
+            encodedBytes.append(new BytesRef(versionParts.buildSuffix));
+        }
+        return new EncodedVersion(
+            encodedBytes.toBytesRef(),
+            true,
+            versionParts.preRelease != null,
+            mainVersionParts[0],
+            mainVersionParts[1],
+            mainVersionParts[2]
+        );
+    }
+
+    private static Integer[] prefixDigitGroupsWithLength(String input, BytesRefBuilder result) {
```

Comment:
That's quite a clever encoding! Nice work! What I like about it is that it allows the decoding to the original values.

Another way to do encoding I was thinking of is to standardize main versions to 3 bytes (1 byte of each component), but standardization would not allow the decoding to the original values.

Reply:
Yes, being able to get back the original values is quite important here, e.g. we need that when needing to display encoded docvalues (e.g. keys for terms aggs where I first encountered that problem). I tried several approaches with encoding main verison parts as bytes (several widths) but they all failed either because bein too restrictive (we also want to allow less or more than 3 main digits, as an extension to strict semver, e.g. "1.0.14.20201027") and the range of verison numbers is in theory unlimited. Thats why in the end I prefer this simpler apporoach which can also be used for numeric parts in the pre-release part of the version string.